### PR TITLE
stm32: Add support for Nucleo-F12ZG development board

### PR DESCRIPTION
**Pull request by TPRoberts**
This port was loosely based off the port for NUCLEO_F413ZH. Information on the development board can be found [here](https://www.st.com/en/evaluation-tools/nucleo-f412zg.html).
Build instructions:
cd port/stm32
make BOARD=NUCLEO_F412ZG

I haven't test every component but been using it on my development board for a quick projects.

### CODE REVIEW

File: ports/stm32/adc.c [hunk 1]
```diff
@@ -135,7 +135,7 @@
 #elif defined(STM32F411xE) || defined(STM32F413xx) || \
     defined(STM32F427xx) || defined(STM32F429xx) || \
     defined(STM32F437xx) || defined(STM32F439xx) || \
-    defined(STM32F446xx)
+    defined(STM32F446xx) || defined(STM32F412Zx)
```
Comment (dpgeorge) [inline ports/stm32/adc.c:138]: Would be good to put this in sorted order between f411 and f413

File: ports/stm32/boards/stm32f412_af.csv [hunk 2]
```diff
@@ -0,0 +1,116 @@
+Port,,AF0,AF1,AF2,AF3,AF4,AF5,AF6,AF7,AF8,AF9,AF10,AF11,AF12,AF13,AF14,AF15,
```
Comment (dpgeorge) [inline ports/stm32/boards/stm32f412_af.csv:1]: Is this AF file exactly the same as an existing one, or is it F413-specific?

File: ports/stm32/flashbdev.c [hunk 3]
```diff
@@ -53,7 +53,7 @@
 #define FLASH_MEM_SEG2_NUM_BLOCKS (128) // sector 11: 128k
 #endif
 
-#elif defined(STM32F401xE) || defined(STM32F411xE) || defined(STM32F446xx)
+#elif defined(STM32F401xE) || defined(STM32F411xE) || defined(STM32F446xx) || defined(STM32F412Zx)
```
Comment (dpgeorge) [inline ports/stm32/flashbdev.c:56]: please put the F412 in sorted order between F411 and F446

File: ports/stm32/modmachine.c [hunk 4]
```diff
@@ -74,6 +74,13 @@
 #define RCC_SR_BORRSTF  RCC_RSR_BORRSTF
 #define RCC_SR_PINRSTF  RCC_RSR_PINRSTF
 #define RCC_SR_RMVF     RCC_RSR_RMVF
+#elif defined(STM32F412Zx)
+#define RCC_SR          CSR
+#define RCC_SR_IWDGRSTF RCC_CSR_IWDGRSTF
+#define RCC_SR_WWDGRSTF RCC_CSR_WWDGRSTF
+#define RCC_SR_PORRSTF  RCC_CSR_PORRSTF
+#define RCC_SR_PINRSTF  RCC_CSR_PINRSTF
+#define RCC_SR_RMVF     RCC_CSR_RMVF
```
Comment (dpgeorge) [inline ports/stm32/modmachine.c:83]: is this addition needed at all?  looks like it uses the default defines below

**Comment by dpgeorge** [comment]
Thanks for the contribution.  To maintain a good git history, can you please split the commit here into 2 commits: one which adds support for F412 MCUs in general (changes to adc.c, flash.c, addition of .ld and af.csv files), and then one commit which adds the board config files for the new Nucleo board.

File: ports/stm32/boards/stm32f412_af.csv (hunk 2 above)
Comment (TPRoberts) [reply ports/stm32/boards/stm32f412_af.csv:1]: I can't remember exactly as it was months ago when I added this support but when I created it I remember there to be differences with any other file and manually working through the Reference Manual. I also just did a diff with some chips in the same family and there was differences, so believe it is needed. Thanks.

**Comment by dpgeorge** [comment]
Thanks for updating.  I checked that the af.csv file is different to others, so it should be kept.

Rebased and merged in d7399679ded88a2f8b64f69f487bc52d648ca6fd and 463c0fb2f4fa5dfd79598c4d8681a007704c2ee1 with some minor code formatting applied to `mpconfigboard.h`.
